grudge/symbolic/operators.py

```diff
@@ -651,20 +658,53 @@ def matrix(self, afgrp, volgrp, dtype):
                     afgrp.nunit_dofs),
                 dtype=dtype)
 
-        from modepy.tools import UNIT_VERTICES
         import modepy as mp
-        for iface, fvi in enumerate(
-                volgrp.mesh_el_group.face_vertex_indices()):
-            face_vertices = UNIT_VERTICES[volgrp.dim][np.array(fvi)].T
-            matrix[:, iface, :] = mp.nodal_face_mass_matrix(
-                    volgrp.basis(), volgrp.unit_nodes, afgrp.unit_nodes,
-                    volgrp.order,
-                    face_vertices)
-
-        # np.set_printoptions(linewidth=200, precision=3)
-        # matrix[np.abs(matrix) < 1e-13] = 0
-        # print(matrix)
-        # 1/0
+        from meshmode.discretization import ElementGroupWithBasis
+        from meshmode.discretization.poly_element import \
+            QuadratureSimplexElementGroup
+
+        n = volgrp.order
+        m = afgrp.order
+        vol_basis = volgrp.basis_obj()
+        faces = mp.faces_for_shape(volgrp.shape)
+
+        for iface, face in enumerate(faces):
+            # If the face group is defined on a higher-order
+            # quadrature grid, use the underlying quadrature rule
+            if isinstance(afgrp, QuadratureSimplexElementGroup):
+                face_quadrature = afgrp._quadrature_rule()
+            else:
+                # NOTE: This handles the general case where
+                # volume and surface quadrature rules may have different
+                # integration orders
+                face_quadrature = mp.quadrature_for_space(
+                    mp.space_for_shape(face, 2*max(n, m)),
+                    face
+                )
+
+            assert face_quadrature.exact_to >= n + m
```

Comment:
@inducer I have a couple questions that's puzzling me and I want to make sure I understand  what's going on.

1. When overintegrating, why is `volgrp` from the native (not overintegrated) discretization? Is it to ensure you have access to the volume basis functions so you can sample at the higher-order quadrature grid?

2. What should the accuracy check be when `afgrp` is coming from a higher-order quadrature grid? I think this assert here:
```python
assert face_quadrautre.exact_to >= n + m
```
Makes sense for all typical cases (no overintegration). No questions there. However,
for example in [this mirgecom test](https://github.com/illinois-ceesd/mirgecom/blob/2a39aa880ff883774f9646e23746c682c151fd80/test/test_diffusion.py#L271), `volgrp.order` is 2, but afgrp.order is 6. The corresponding quadrature rule associated with `afgrp` has an accuracy such that it exactly integrates polynomials of order 6 (based on checking `quad.exact_to`.

So that will fail the assert (at least in 3D). I guess my question is: can't we just safely use the quadrature rule associated with the `QuadratureSimplexElementGroup` and `assert quad.exact_to >= m`? (Otherwise why do we carry around quadrature rules in the first place?)

Reply:
> Is it to ensure you have access to the volume basis functions so you can sample at the higher-order quadrature grid?

Yes! Those are the basis functions you need to test against after all.

> can't we just safely use the quadrature rule associated with the `QuadratureSimplexElementGroup` and `assert quad.exact_to >= m`? (Otherwise why do we carry around quadrature rules in the first place?)

That works for me. Possibly, an argument could be made for `quad.exact_to >= 2*m`, but I'm OK with whichever you choose.

Side point: Don't use `assert` for validating things that are under user control. (those should be `if: raise`). `assert` is for things you are able to prove true based on assumptions the code makes and the algorithms you're using. (It's possible that you'll spot bits of the code that don't follow that guideline. If so, feel free to change them.)